feat: enable HTTPRoute for Gateway API

Kyrremann opened the pull request:
We are moving to Gateway API with HTTPRoute instead of using regular Ingress resources. It would be nice if this chart would allow us to make the two resources needed so we don't have to make them manually before installing the chart.

From https://gateway-api.sigs.k8s.io/

When using the Gateway API to manage ingress traffic, the Gateway resource defines a point of access at which traffic can be routed across multiple contexts -- for example, from outside the cluster to inside the cluster (north/south traffic).

<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->



<!-- Please keep an empty line above the dashes. -->

Kyrremann committed:
feat: enable HTTPRoute for Gateway API

From https://gateway-api.sigs.k8s.io/

> When using the Gateway API to manage ingress traffic, the Gateway resource defines a point of access at which traffic can be routed across multiple contexts -- for example, from outside the cluster to inside the cluster (north/south traffic).

hussein-awala commented on chart/values.yaml:
```diff
@@ -220,6 +220,31 @@ ingress:
+    # The Gateway API configuration
+    gateway:
+      name: your-gateway
+      namespace: namespace-for-your-gateway
```
You can keep them empty, and check in the template if they are provided, and fail if they are not (at least name, and maybe use chart namespace as namespace for gateway). Otherwise, users will try to use it with default values, and it will fail.

Kyrremann replied:
I'll leave them both empty. My guess is that k8s admins will not have the gateway running in the same namespace as Airflow.

hussein-awala commented on chart/values.yaml:
```diff
@@ -220,6 +220,31 @@ ingress:
+    # The Gateway API configuration
+    gateway:
+      name: your-gateway
+      namespace: namespace-for-your-gateway
```
same here

hussein-awala reviewed (commented):
Could you add test cases for these two templates, and add the new values to the values schema file?

jedcunningham commented on chart/values.yaml:
```diff
@@ -220,6 +220,31 @@ ingress:
+# Enable the creation of a Kubernetes HTTPRoute and HealthCheckPolicy
+httpRoute:
+  web:
+    # Enable web ingress resource
```
```suggestion
    # Enable web HTTPRoute resource
```

nit

jedcunningham commented on chart/templates/webserver/webserver-httproute.yaml:
```diff
@@ -0,0 +1,82 @@
+  rules:
+    - backendRefs:
+        - name: {{ $fullname }}-webserver
+          port: 8080
```
Can this use a named port (`airflow-ui`) instead? If not, it needs to use `.Values.ports.airflowUI` instead.

Kyrremann replied:
As far as I can see from the documentation it needs to be the port number.
https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io/v1beta1.PortNumber

jedcunningham commented on chart/templates/flower/flower-httproute.yaml:
```diff
@@ -0,0 +1,82 @@
+  rules:
+    - backendRefs:
+        - name: {{ $fullname }}-flower
+          port: 8080
```
Same here, plus flower uses 5555 by default?

Kyrremann committed:
fix: wrong resource name in documentation

Co-authored-by: Jed Cunningham <66968678+jedcunningham@users.noreply.github.com>

Kyrremann committed:
feat: use port variables

Kyrremann committed:
fix: there is no host configuration for httproute

Kyrremann committed:
feat: use empty values and required in template

Kyrremann committed:
feat: simple test of httproute for web

Kyrremann committed:
fix: fixed failing tests

Kyrremann committed:
feat: needs to add Gateway API CRDs for helm tests

Kyrremann committed:
feat: same httproute for flower

Kyrremann committed:
feat: added httproute to schema

jedcunningham commented on helm_tests/webserver/test_httproute_flower.py:
```diff
@@ -0,0 +1,97 @@
+                                          "gateway": {"name": "test-gateway", "namespace": "test"}}}},
+            show_only=["templates/flower/flower-httproute.yaml"],
+        )  # checks that no validation exception is raised
+
```
```suggestion
```

I'm not sure we need this test when we have the next one. Each test isn't that big, but in total they add up.

jedcunningham commented on helm_tests/webserver/test_httproute_flower.py:
```diff
@@ -0,0 +1,97 @@
+                                          "gateway": {"name": "test-gateway", "namespace": "test"}}}},
+            show_only=["templates/flower/flower-httproute.yaml"],
+        )
+        assert not jmespath.search("spec.hostnames", docs[0])
```
Should we have a test that hostnames do get added as well?

Kyrremann replied:
It's probably better to only have a test that checks if they are added. Do we really care about them not getting added if not specified?

jedcunningham replied:
Yeah, let's just rewrite this test to cover them being added.

jedcunningham commented on helm_tests/webserver/test_httproute_flower.py:
```diff
@@ -0,0 +1,97 @@
+        if value is not None:
+            values["httpRoute"]["flower"]["enabled"] = value
+        docs = render_chart(values=values, show_only=["templates/flower/flower-httproute.yaml"])
+        assert expected == (2 == len(docs))
```
It might be better to actually assert the count directly. If you got 1 doc and checked "false" it'd still match.

jedcunningham commented on chart/templates/flower/flower-httproute.yaml:
```diff
@@ -0,0 +1,82 @@
+kind: HTTPRoute
+apiVersion: gateway.networking.k8s.io/v1beta1
+metadata:
+  name: {{ $fullname }}-httproute
```
Shouldn't this name be specific to flower? What if someone wanted to enable both?

Kyrremann replied:
Good catch!

jedcunningham commented on helm_tests/webserver/test_httproute_flower.py:
```diff
@@ -0,0 +1,97 @@
+            show_only=["templates/flower/flower-httproute.yaml"],
+        )
+        assert "test_label" in jmespath.search("metadata.labels", docs[0])
+        assert jmespath.search("metadata.labels", docs[0])["test_label"] == "test_label_value"
```
Shouldn't we check the HealthCheckPolicy also?

Kyrremann committed:
fix: proper name for flower httproute

Kyrremann committed:
fix: removed unnecessary test

Kyrremann committed:
fix: better formatting of code

Kyrremann committed:
feat: also check healthcheckpolicy

Kyrremann committed:
fix: no need to specify resource name in name

Kyrremann committed:
feat: healthcheckpolicy service is set

Kyrremann committed:
fix: expected value is number of resources created

Kyrremann committed:
fix: had way too many commas

Kyrremann committed:
fix: wrong filename for webserver

Kyrremann committed:
Merge branch 'main' into gateway_api

Kyrremann committed:
fix: added items keys after linting

> Found array type definitions with too loose a definition. Make sure the object has the items key.

Kyrremann committed:
fix: ruff formating

Kyrremann committed:
fix: removed noqa tag
